tests/adaptive_mh_01.cc

```diff
@@ -0,0 +1,117 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2019 by the SampleFlow authors.
+//
+// This file is part of the SampleFlow library.
+//
+// The SampleFlow library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of SampleFlow.
+//
+// ---------------------------------------------------------------------
+
+
+// Like the covariance_matrix_{10,11} tests, but after the first 500
+// samples, use the actual covariance matrix as the proposal
+// distribution.
+
+
+#include <iostream>
+#include <fstream>
+#include <eigen3/Eigen/Dense>
+
+#include <sampleflow/producers/metropolis_hastings.h>
+#include <sampleflow/consumers/covariance_matrix.h>
+#include <sampleflow/consumers/mean_value.h>
+#include <sampleflow/consumers/count_samples.h>
+
+using SampleType = Eigen::Vector2d;
+
+
+double log_likelihood (const SampleType &x)
+{
+  const SampleType mu = {1,2};
+  const SampleType y = x-mu;
+  Eigen::Matrix2d C;
+  C << 1, 0.1,
+  0.1, 1;
+  return -0.5 * (y.transpose()*(C.inverse()*y))(0,0);
+}
+
+
+
+// Use the non-adaptive proposal distribution from the _10 test for
+// the first 500 samples
+std::pair<SampleType,double> perturb_simple (const SampleType &x)
+{
+  static std::mt19937 rng;
+  const double delta = 0.1;
+  std::uniform_real_distribution<double> distribution(-delta,delta);
+
+  SampleType y = x;
+  for (unsigned int i=0; i<y.size(); ++i)
+    y[i] += distribution(rng);
```

Comment:
My understanding of Eigen is that for vectors the [ ] syntax is valid, but it is preferred to use the ( ) syntax. I think this is for consistency because I don't think the [ ] syntax works for matrices. If you have a matrix object where one dimension is 1, then Eigen understands that it is a vector and you have x[1]=x(1)=x(1,0).

Reply:
I'd accept a patch :-)